text/3536-unsized-thin-pointers.md

```diff
@@ -0,0 +1,291 @@
+custom pointer metadata from `&self`.
+
+Implementing `DynSized` does not affect alignment, so the questions of how to
+handle unknown alignments of RFC 1861 `extern type` DSTs do not apply.
+
+In current Rust, a DST used as a `struct` field must be the final field of the
+`struct`. This restriction remains unchanged, as the offsets of any fields after
+a DST would be impossible to compute statically.
+- This also implies that any given `struct` may have at most one field that
+  implements `DynSized`.
+
+A `struct` with a field that implements `DynSized` will also implicitly
+implement `DynSized`. The implicit implementation of `DynSized` computes the
+size of the struct up until the `DynSized` field, and then adds the result of
+calling `DynSized::size_of_val()` on the final field.
+- This implies it's not permitted to manually `impl DynSize` for a type that
+  contains a field that implements `DynSize`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Mutability of value sizes
+
+If the size of a value is stored in the value itself, then that implies it can
+change at runtime.
+
+```rust
+struct MutableSize { size: usize }
+unsafe impl core::mem::DynSized for MutableSize {
+	fn size_of_val(&self) -> usize { self.size }
+}
+
+let mut v = MutableSize { size: 8 };
+println!("{:?}", core::mem::size_of_val(&v)); // prints "8"
+v.size = 16;
+println!("{:?}", core::mem::size_of_val(&v)); // prints "16"
+```
+
+There may be existing code that assumes `size_of_val()` is constant for a given
+value, which is true in today's Rust due to the nature of fat pointers, but
+would no longer be true if `size_of_val()` is truly dynamic.
+
+Alternatively, the API contract for `DynSized` implementations could require
+that the result of `size_of_val()` not change for the lifetime of the allocated
+object. This would likely be true for nearly all interesting use cases, and
+would let `DynSized` values be stored in a `Box`.
+
+## Compatibility with existing fat-pointer DSTs
+
+It may be desirable for certain existing stabilized DSTs to implement
+`DynSized` -- for example, it is a natural fit for the planned redefinition of
+[`&core::ffi::CStr`][cstr] as a thin pointer.
+
+[cstr]: https://doc.rust-lang.org/core/ffi/struct.CStr.html
+
+Such a change to existing types might be backwards-incompatible for code that
+embeds those types as a `struct` field, because it would change the reference
+layout. For example, the following code compiles in stable Rust v1.73 but would
+be a compilation error if `&CStr` does not have the same layout as `&[u8]`.
+
+```rust
+struct ContainsCStr {
+	cstr: core::ffi::CStr,
+}
+impl ContainsCStr {
+	fn as_bytes(&self) -> &[u8] {
+		unsafe { core::mem::transmute(self) }
+	}
+}
+```
+
+The above incompatibility of a redefined `&CStr` exists regardless of this RFC,
+but it's worth noting that implementing `DynSized` would be a backwards
+incompatible change for existing DSTs.
+
+# Rationale and alternatives
```

Comment:
Have you considered just having everything automatically implement `DynSized` except `extern` types and structs ending with them directly or indirectly except through `UnsafeCell<T>` (thanks @CAD97)?

In this world:

- `Sized`: `std::mem::size_of::<T>()` works and `DynSized` returns the result of that.
- `DynSized`: `std::mem::size_of_val(&t)` works and delegates to `<T as DynSized>::size_of_val(&t)`
- Extern types must manually implement `DynSized` if applicable.
- `UnsafeCell<T>` doesn't proxy through extern type implementations as it depends on the outer type how it should be accessed. For instance, mutexes could return `mutex.lock().unwrap_or_else(|e| e.into_inner()).size_of_val()` and read-write locks `rwlock.read().unwrap_or_else(|e| e.into_inner()).size_of_val()`.